src/util.js

```diff
@@ -52,29 +58,184 @@ export const SPECIAL_TOKENS = {
+ * new objects with the CustomValue API
+ * (.i.e `{ _custom: { ... } }`)
+ */
+class EncodeCache {
+  constructor () {
+    this.map = new Map()
+  }
+
+  /**
+   * Returns a result unique to each input data
+   * @param {*} data Input data
+   * @param {*} factory Function used to create the unique result
+   */
+  cache (data, factory) {
+    const cached = this.map.get(data)
+    if (cached) {
+      return cached
+    } else {
+      const result = factory(data)
+      this.map.set(data, result)
+      return result
+    }
+  }
+
+  clear () {
+    this.map.clear()
+  }
+}
+
+const encodeCache = new EncodeCache()
+
 export function stringify (data) {
+  // Create a fresh cache for each serialization
+  encodeCache.clear()
   return CircularJSON.stringify(data, replacer)
 }
 
 function replacer (key) {
   const val = this[key]
-  if (val === undefined) {
+  const type = typeof val
+  if (type === 'undefined') {
     return UNDEFINED
   } else if (val === Infinity) {
     return INFINITY
   } else if (val === -Infinity) {
     return NEGATIVE_INFINITY
+  } else if (type === 'function') {
+    return getCustomFunctionDetails(val)
+  } else if (val !== null && type === 'object') {
+    if (val instanceof Map) {
+      return encodeCache.cache(val, () => getCustomMapDetails(val))
+    } else if (val instanceof Set) {
+      return encodeCache.cache(val, () => getCustomSetDetails(val))
+    } else if (val instanceof RegExp) {
+      // special handling of native type
+      return `[native RegExp ${val.toString()}]`
+    } else if (val instanceof Date) {
+      return `[native Date ${val.toString()}]`
+    } else if (val.state && val._vm) {
+      return encodeCache.cache(val, () => getCustomStoreDetails(val))
+    } else if (val.constructor && val.constructor.name === 'VueRouter') {
+      return encodeCache.cache(val, () => getCustomRouterDetails(val))
+    } else if (val._isVue) {
+      return encodeCache.cache(val, () => getCustomInstanceDetails(val))
+    } else if (typeof val.render === 'function') {
+      return encodeCache.cache(val, () => getCustomComponentDefinitionDetails(val))
+    }
   } else if (Number.isNaN(val)) {
     return NAN
-  } else if (val instanceof RegExp) {
-    // special handling of native type
-    return `[native RegExp ${val.toString()}]`
-  } else if (val instanceof Date) {
-    return `[native Date ${val.toString()}]`
-  } else if (val && val._isVue) {
-    return getCustomInstanceDetails(val)
+  }
+  return sanitize(val)
+}
+
+export function getCustomMapDetails (val) {
+  const list = []
+  val.forEach(
+    (value, key) => list.push({
+      key,
+      value
+    })
+  )
+  return {
+    _custom: {
+      type: 'map',
+      display: 'Map',
+      value: list,
+      readOnly: true,
+      fields: {
+        abstract: true
+      }
+    }
+  }
+}
+
+export function reviveMap (val) {
+  const result = new Map()
+  const list = val._custom.value
+  for (let i = 0; i < list.length; i++) {
```

Comment:
Perhaps you can use `forEach` instead? 
```
list.forEach(({ key, value}) =>
  result.set(key, reviver(null, value)
)
```

Reply:
Simple loop without anonymous function call is faster imho. Also made a test case: https://jsperf.com/map-loop-vs-foreach (for what it's worth :upside_down_face:)